org-roam.el

```diff
@@ -231,15 +252,55 @@ If FILE is not specified, use the current buffer's file-path."
          (f-descendant-of-p (file-truename path)
                             (file-truename org-roam-directory))))))
 
-(defun org-roam--list-files (dir)
-  "Return all Org-roam files located within DIR, at any nesting level.
-Ignores hidden files and directories."
+(defun org-roam--shell-command-files (cmd)
+  "Run CMD in the shell and return a list of files. If no files are found, an empty list is returned."
+  (seq-filter #'s-present? (split-string (shell-command-to-string cmd) "\n")))
+
+(defun org-roam--list-files-search-globs (exts)
+  "Given EXTS, return a list of search globs.
+E.g. (\".org\") => (\"*.org\" \"*.org.gpg\")"
+  (append
+   (mapcar (lambda (ext) (s-wrap (concat "*." ext) "\"")) exts)
+   (mapcar (lambda (ext) (s-wrap (concat "*." ext ".gpg") "\"")) exts)))
+
+(defun org-roam--list-files-rg (dir)
+  "Return all Org-roam files located recursively within DIR, using ripgrep."
+  (let* ((globs (org-roam--list-files-search-globs org-roam-file-extensions))
+         (executable (executable-find "rg"))
+         (command (s-join " " `(,executable ,dir "--files"
+                            ,@(mapcar (lambda (glob) (concat "-g " glob)) globs)))))
+    (if executable
+        (org-roam--shell-command-files command)
+      (lwarn '(org-roam) :error
+             "Cannot find \"rg\" executable. Make sure it is available, or tweak `org-roam-find-file-functions'"))))
+
+(defun org-roam--list-files-find (dir)
+  "Return all Org-roam files located recursively within DIR, using find."
+  (let* ((globs (org-roam--list-files-search-globs org-roam-file-extensions))
+         (executable (executable-find "find"))
+         (command (s-join " " `(,executable ,dir "-type f \\("
+                                            ,(s-join " -o " (mapcar (lambda (glob) (concat "-name " glob)) globs)) "\\)"))))
+    (if executable
+        (org-roam--shell-command-files command)
+      (lwarn '(org-roam) :error
+             "Cannot find \"find\" executable. Make sure it is available, or tweak `org-roam-find-file-functions'"))))
+
+(defun org-roam--list-files-elisp (dir)
+  "Return all Org-roam files located recursively within DIR, using elisp."
   (let ((regex (concat "\\.\\(?:"(mapconcat #'regexp-quote org-roam-file-extensions "\\|" )"\\)\\(?:\\.gpg\\)?\\'"))
         result)
     (dolist (file (directory-files-recursively dir regex) result)
       (when (and (file-readable-p file) (org-roam--org-file-p file))
         (push file result)))))
 
+(defun org-roam--list-files (dir)
+  "Return all Org-roam files located recursively within DIR.
+It will use external shell functions if defined in `org-roam-list-files-functions'."
+  (let ((valid-command (seq-find (lambda (s) (executable-find (symbol-name s))) org-roam-list-files-functions)))
+    (if valid-command
+        (funcall (intern (concat "org-roam--list-files-" (symbol-name valid-command))) dir)
+      (org-roam--list-files-elisp dir))))
```

Comment:
```suggestion
(defun org-roam--list-files (dir)
  "Return all Org-roam files located recursively within DIR.
Use external shell functions if defined in `org-roam-list-files-functions'."
  (let* ((found (seq-find (lambda (s) (executable-find (symbol-name s))) org-roam-list-files-functions))
         ((fn (intern (concat "org-roam--list-files-" (symbol-name found))))))
    (if (not found)
        (org-roam--list-files-elisp dir)
      (unless (fboundp fn) (user-error "%s is not an implemented search method."))
      (funcall fn dir))))
```

We need to account for the case where the symbol is a valid executable, but does not have an implemented search function.

Reply:
Another potential downside to the approach of using a symbol to map to a function name is that the user cannot specify the path to the executable. One possible solution is to accept either a symbol or consed pair. e.g.
```emacs-lisp
(setq org-roam-find-file-commands '(find (rg . "/path/to/rg")))
```
Then you could check the type in the lambda passed to seq-find. If it's a pair, grab the cdr and `excutable-find` that. If it's a symbol, use the strategy you've already implemented.